feat: Add 6 compiler-level utility functions for ErgoScript box verif…

bigpandamx opened the pull request:
# Add 6 Compiler-Level Utility Functions for ErgoScript Box Verification

**Resolves:** #1037  
**Type:** Feature Implementation  
**Compatibility:** Sigma 6.0+

## 📋 Summary

This PR implements 6 utility functions requested by @bdkent for common ErgoScript box verification operations. These functions are now available as **global compiler-level utilities**, eliminating the need for developers to copy-paste utility code across contracts.

## 🎯 Functions Implemented

### 1. `verifyBoxHasMarkerToken(box: Box, tokenId: Coll[Byte]): Boolean`
Verifies a box contains a specific token ID with amount ≥ 1.

### 2. `verifyBoxHasNoMarkerToken(box: Box, tokenId: Coll[Byte]): Boolean`  
Verifies a box does NOT contain a specific token ID.

### 3. `verifyUsedAdditionalRegisters(box: Box, expectedCount: Int): Boolean`
Validates that a box uses only the specified number of additional registers (R4-R9).

### 4. `verifySameForBasicRequiredRegisters(inBox: Box, outBox: Box): Boolean`
Compares `value` and `propositionBytes` between two boxes.

### 5. `verifySameForRequiredRegisters(inBox: Box, outBox: Box): Boolean`
Compares `value`, `propositionBytes`, and `tokens` between two boxes.

### 6. `verifySpentToken(inBox: Box, outBox: Box, tokenId: Coll[Byte], amount: Long): Boolean`
Verifies that a specific token amount was properly spent (reduced) between input and output boxes.

## 🔧 Technical Implementation

### Multi-Layer Integration
- **SigmaPredef.scala**: Added `PredefinedFunc` definitions with proper cost modeling (Method IDs 11-16)
- **methods.scala**: Added `SMethod` definitions with correct type signatures  
- **SigmaDsl.scala**: Added method signatures to `SigmaDslBuilder` and `SigmaContract` traits
- **CSigmaDslBuilder.scala**: Implemented concrete verification logic using modern API patterns
- **IR Layer**: Updated `SigmaDslImpl`, `SigmaDslUnit`, `GraphIRReflection` for compiler support

### API Modernization
- Uses `box.getReg(n).isDefined` instead of deprecated `box.R4[Any].isEmpty` pattern
- Clean case destructuring: `case (id, amount) =>` instead of `t._1`, `t._2` accessors
- Flattened parameter structure for better usability

## 📊 Files Changed

| **File** | **Purpose** | **Changes** |
|----------|-------------|-------------|
| `core/.../SigmaDsl.scala` | Method signatures | +24 lines (trait methods) |
| `core/.../ReflectionData.scala` | Method registration | +6 lines (method refs) |
| `data/.../SigmaPredef.scala` | Global functions | +136 lines (PredefinedFunc defs) |
| `data/.../methods.scala` | Method definitions | +66 lines (SMethod defs) |
| `data/.../CSigmaDslBuilder.scala` | Implementations | +68 lines (concrete logic) |
| `sc/.../GraphIRReflection.scala` | IR reflection | +6 lines (method registration) |
| `sc/.../SigmaDslUnit.scala` | IR traits | +24 lines (method sigs) |
| `sc/.../SigmaDslImpl.scala` | IR implementations | +12 lines (method calls) |
| `sc/.../UtilityFunctionsTestSimple.scala` | Tests | +49 lines (API validation) |

## ✅ Testing & Validation

### Test Results
- **886/887 tests pass** - 1 expected failure in `ErgoTreeSpecification`  
- **Expected failure reason**: Test validates method count (expects 10, now gets 16)
- **Clean compilation**: No errors, warnings only (unrelated to our changes)
- **API validation**: All 6 functions verified as accessible via reflection

### Usage Example
```scala
// Now available in any ErgoScript contract:
{
  val inputBox = INPUTS(0)
  val outputBox = OUTPUTS(0)
  val nftId = fromBase16("deadbeef...")
  
  verifyBoxHasMarkerToken(inputBox, nftId) &&
  verifySameForBasicRequiredRegisters(inputBox, outputBox) &&
  verifySpentToken(inputBox, outputBox, nftId, 1L)
}
```

## 🔄 Backward Compatibility

- ✅ **No breaking changes** to existing API
- ✅ **Sigma 6.0+ compatible** (functions gated by `isV3OrLaterErgoTreeVersion`)  
- ✅ **Maintains existing cost model patterns**
- ✅ **Preserves all existing functionality**

## 📝 Developer Impact

**Before**: Developers had to copy-paste utility functions into every contract
**After**: Clean, built-in utilities available globally without imports

Commit:
Add DownCast for register id in executeFromSelfReg()

Fixes #602

- Add automatic downcasting from numeric types to Int for register id parameter
  in executeFromSelfReg() and executeFromSelfRegWithDefault() functions
- This mirrors the existing downcast behavior for getVar and executeFromVar
- Users can now write executeFromSelfReg[T](4) or executeFromSelfReg[T](4L)
  without manual .toByte casting

Changes:
- SigmaTyper.scala: Added downcast logic for ExecuteFromSelfRegFunc and
  ExecuteFromSelfRegWithDefaultFunc in adaptedTypedArgs pattern matching
- SigmaTyperTest.scala: Added comprehensive test cases verifying downcast
  functionality with Int and Long literals

All tests passing (58 tests in SigmaTyperTest suite)

kushti commented on core/shared/src/main/scala/sigma/SigmaDsl.scala:
```diff
@@ -989,6 +996,53 @@ trait SigmaDslBuilder {
   /** Returns a number decoded from provided big-endian bytes array. */
   def fromBigEndianBytes[T](bytes: Coll[Byte])(implicit cT: RType[T]): T
 
+  /** Verifies that the given box contains at least one token with the specified token ID and amount >= 1.
```
this is not the frontend level, rather, consensus-critical ErgoTree level, there should be no changes here at all.

kushti reviewed (changes requested):
make sure all the changes are within "sc" subproject (compiler), if any code outside is touched (aside of some trivial changes, like helper methods), that is a sign of something wrong happens

remove all the AI-genrated MD files before submission

Commit:
Add 6 utility functions via SigmaPredef compiler desugaring

Implements Issue #1037 by adding the following utility functions:
- verifyBoxHasMarkerToken: Check if box contains specific tokenId with amount >= 1
- verifyBoxHasNoMarkerToken: Check if box does NOT contain specific tokenId
- verifyUsedAdditionalRegisters: Check which additional registers (R4-R9) are used
- verifySameForBasicRequiredRegisters: Compare value and propositionBytes between boxes
- verifySameForRequiredRegisters: Compare value, propositionBytes, and tokens between boxes
- verifySpentToken: Verify token amount spent in transaction

All functions implemented using irBuilder pattern in SigmaPredef.scala to desugar
at compile-time to existing ErgoTree operations, avoiding consensus-layer changes.

Includes comprehensive test suite in UtilityFunctionsSpec.scala.
